lib/spack/spack/solver/concretize.lp

```diff
@@ -461,13 +461,58 @@ error(100, "Attempted to use external for '{0}' which does not satisfy any confi
 % Config required semantics
 %-----------------------------------------------------------------------------
 
-activate_requirement_rules(Package) :- attr("node", Package).
-activate_requirement_rules(Package) :- attr("virtual_node", Package).
+activate_requirement(Package, X) :-
+  attr("node", Package),
+  requirement_group(Package, X),
+  not requirement_conditional(Package, X).
+
+activate_requirement(Package, X) :-
+  attr("virtual_node", Package),
+  requirement_group(Package, X),
+  not requirement_conditional(Package, X).
+
+activate_requirement(Package, X) :-
+  attr("node", Package),
+  1 { condition_holds(Y) : requirement_group_member(Y, Package, X) },
+  requirement_group(Package, X),
+  requirement_conditional(Package, X).
```

Comment:
(abstract design issue, more of a discussion prompt than a request) I find this confusing:

* My assumption is that `activate_requirement(Package, X)` means "the requirement group `X` for `Package` must be applied" (i.e. that one of the members must hold)
* This seems to say that in order for the requirement group X to be mandatory, that at least one of its members must hold

To me, the members of the requirement group are things that have to be applied only if we decide the requirement group is mandatory. I figured that would be achievable by replacing

```
activate_requirement_rules(Package) :- attr("node", Package).
```

with

```
activate_requirement_group(Package, X) :- 
  attr("node", Package),
  requirement_group(Package, X),
  requirement_condition(Z, X),
  condition_holds(Z).
```

the difference being that the `requirement_condition` Z is not a member of the `requirement_group` (well, it could be, but that would seem useless to me, like `require: +foo when: +foo`).

Reply:
See https://github.com/spack/spack/pull/36286#discussion_r1165307712 For a conditional requirement all the conditions are equal and map to the `when=` argument. So saying:
```prolog
1 { condition_holds(Y) : requirement_group_member(Y, Package, X) },
```
means that the `when=` holds and we should consider the requirement.